README.md
# calc_lat_long

A package to calculate distance between two cordinate points Latitude and Longitude by Miles or Nautical Miles or Kilometers.

## Getting Started

1. Add the latest version of package to your pubspec.yaml (and run `dart pub get`);

```yaml
depencencies:
  calc_lat_long: ^1.0.7
```

## Example

There are a number of properties that you can modify:

- Latitude1
- Longitude1
- Latitude2
- Longitude2

- Enum UnitLength.km [mi, km, nm] (miles, kilometers, nautical miles)

<hr>

```dart
Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Padding(
        padding: EdgeInsets.all(16),
        child: ListView(
          keyboardDismissBehavior: ScrollViewKeyboardDismissBehavior.onDrag,
          children: <Widget>[
            TextField(
              controller: latitude1,
              decoration: InputDecoration(labelText: "Latitude 1"),
            ),
            TextField(
              controller: longitude1,
              decoration: InputDecoration(labelText: "Longitude 1"),
            ),
            TextField(
              controller: latitude2,
              decoration: InputDecoration(labelText: "Latitude 2"),
            ),
            TextField(
              controller: longitude2,
              decoration: InputDecoration(labelText: "Longitude 2"),
            ),
            DropdownButton<UnitLength>(
                items: _unitLength.map((UnitLength dropDownStringItem) {
                  return DropdownMenuItem<UnitLength>(
                    value: dropDownStringItem,
                    child: Text(dropDownStringItem.toString()),
                  );
                }).toList(),
                onChanged: (UnitLength newUnitSelected) {
                  _dropDownItemSelected(newUnitSelected);
                  setState(() {
                    _selectedUnit = newUnitSelected;
                  });
                },
                value: _selectedUnit),
            SizedBox(
              height: 15,
            ),
            ElevatedButton(
                onPressed: () {
                  /// Here you must pass your coordinates following by Latitude 1, Latitude 2, Logintude 1, Longitude 2
                  var dist = CalcDistance.distance(
                      num.parse(latitude1.text),
                      num.parse(latitude2.text),
                      num.parse(longitude1.text),
                      num.parse(longitude2.text),
                      _selectedUnit);

                  setState(() {
                    distance = dist;
                  });
                },
                child: Text("Calculate")),
            SizedBox(
              height: 15,
            ),
            Text(distance == 0 ? "" : "Distance: $distance")
          ],
        ),
      )
    );
  }
```


CHANGELOG.md
## [1.0.0]

- Initial release

## [1.0.1]

- Adjusts in README

## [1.0.2]

- Change target version in doc

## [1.0.3]

- Adding information in doc abour Unit of return after calculate

## [1.0.4]

- Adding the same information below in the pubspec

## [1.0.5]

- Upgrading to null safety

## [1.0.6]

- Adjust dart version in pubspec

## [1.0.7]

- Adjust in example to null safety


test/calc_lat_long_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:calc_lat_long/calc_lat_long.dart';

void main() {
  test('Check distance of two coordinates', () {
    final distance = CalcDistance.distance(
        double.parse('41.600097'),
        double.parse('41.60428'),
        double.parse('-93.62959'),
        double.parse('-93.64298'),
        UnitLength.km);
    expect(distance, 1.21);
  });
}


lib/calc_lat_long.dart
library calc_lat_long;

import 'dart:math' as math;

import 'UnitLengthEnum.dart';

export 'UnitLengthEnum.dart';

class CalcDistance {
  static distance(double lat1, double lat2, double lon1, double lon2,
      UnitLength unitLength) {
    if ((lat1 == lat2) && (lon1 == lon2)) {
      return 0;
    } else {
      var radLat1 = math.pi * lat1 / 180;
      var radLat2 = math.pi * lat2 / 180;
      var theta = lon1 - lon2;
      var radTheta = math.pi * theta / 180;
      var dist = math.sin(radLat1) * math.sin(radLat2) +
          math.cos(radLat1) * math.cos(radLat2) * math.cos(radTheta);

      if (dist > 1) {
        dist = 1;
      }

      dist = math.acos(dist);
      dist = dist * 180 / math.pi;
      dist = dist * 60 * 1.1515;

      if (unitLength == UnitLength.km) {
        dist = (dist * 1.609344);
      }

      if (unitLength == UnitLength.nm) {
        dist = (dist * 0.8684);
      }

      return double.parse(dist.toStringAsFixed(2));
    }
  }
}


example/lib/main.dart
import 'package:calc_lat_long/calc_lat_long.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Example calculate two points'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  final TextEditingController latitude1 = TextEditingController();
  final TextEditingController latitude2 = TextEditingController();
  final TextEditingController longitude1 = TextEditingController();
  final TextEditingController longitude2 = TextEditingController();
  var distance = 0.0;
  var _unitLength = [UnitLength.km, UnitLength.mi, UnitLength.nm];
  UnitLength _selectedUnit = UnitLength.km;

  void _dropDownItemSelected(UnitLength novoItem) {
    setState(() {
      _selectedUnit = novoItem;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: Padding(
          padding: EdgeInsets.all(16),
          child: ListView(
            keyboardDismissBehavior: ScrollViewKeyboardDismissBehavior.onDrag,
            children: <Widget>[
              TextField(
                controller: latitude1,
                decoration: InputDecoration(labelText: "Latitude 1"),
              ),
              TextField(
                controller: longitude1,
                decoration: InputDecoration(labelText: "Longitude 1"),
              ),
              TextField(
                controller: latitude2,
                decoration: InputDecoration(labelText: "Latitude 2"),
              ),
              TextField(
                controller: longitude2,
                decoration: InputDecoration(labelText: "Longitude 2"),
              ),
              DropdownButton<UnitLength>(
                  items: _unitLength.map((UnitLength dropDownStringItem) {
                    return DropdownMenuItem<UnitLength>(
                      value: dropDownStringItem,
                      child: Text(dropDownStringItem.toString()),
                    );
                  }).toList(),
                  onChanged: (UnitLength? newUnitSelected) {
                    _dropDownItemSelected(newUnitSelected!);
                    setState(() {
                      _selectedUnit = newUnitSelected;
                    });
                  },
                  value: _selectedUnit),
              SizedBox(
                height: 15,
              ),
              ElevatedButton(
                  onPressed: () {
                    var dist = CalcDistance.distance(
                        double.parse(latitude1.text),
                        double.parse(latitude2.text),
                        double.parse(longitude1.text),
                        double.parse(longitude2.text),
                        _selectedUnit);

                    setState(() {
                      distance = dist;
                    });
                  },
                  child: Text("Calculate")),
              SizedBox(
                height: 15,
              ),
              Text(distance == 0 ? "" : "Distance: $distance")
            ],
          ),
        ));
  }
}
